cds/container/hopscotch_hashmap.h

```diff
@@ -0,0 +1,660 @@
+				std::hash<K> hash_fn;
+				return hash_fn(key) % MAX_SEGMENTS;
+			}
+
+			void resize() {
+				// TODO need to find out if we need one
+			}
+
+		public:
+			static bool const c_isSorted = false; ///< whether the probe set should be ordered
+
+			hopscotch_hashmap() {
+				segments_arys = new Bucket[MAX_SEGMENTS + ADD_RANGE];
+				m_item_counter.reset();
+				BUSY = (KEY*)std::malloc(sizeof(KEY));
+			}
+
+			~hopscotch_hashmap() {
+				std::free(BUSY);
+				std::free(segments_arys);
+			}
+
+			bool empty() {
+				return size() == 0;
+			}
+
+			int size() {
+				return m_item_counter;
+				/*
+				unsigned int counter = 0;
+				const unsigned int num_elm(MAX_SEGMENTS + ADD_RANGE);
+				for (unsigned int iElm = 0; iElm < num_elm; ++iElm) {
+					if (Bucket::_empty_hop_info != (segments_arys + iElm)->_hop_info) {
+						counter += __popcnt((segments_arys + iElm)->_hop_info);
+					}
+				}
+				return counter;
+				*/
+			}
+
+			/// Checks whether the map contains \p key
+			/**
+			The function searches the item with key equal to \p key
+			and returns \p true if it is found, and \p false otherwise.
+			*/
+			template <typename K>
+			bool contains(K const& key)
+			{		
+				return find_with(key, [=](K const& one, K const& two) { return one != two; }, [](mapped_type&) {});
+			}
+
+			/// Checks whether the map contains \p key using \p pred predicate for searching
+			/**
+			The function is similar to <tt>contains( key )</tt> but \p pred is used for key comparing.
+			\p Less functor has the interface like \p std::less.
+			\p Less must imply the same element order as the comparator used for building the map.
+			*/
+			template <typename K, typename Predicate>
+			bool contains(K const& key, Predicate pred)
+			{
+				return find_with(key, pred, [](mapped_type&) {});
+			}
+
+			/// Find the key \p key
+			/** \anchor cds_nonintrusive_CuckooMap_find_func
+
+			The function searches the item with key equal to \p key and calls the functor \p f for item found.
+			The interface of \p Func functor is:
+			\code
+			struct functor {
+			void operator()( value_type& item );
+			};
+			\endcode
+			where \p item is the item found.
+
+			The functor may change \p item.second.
+
+			The function returns \p true if \p key is found, \p false otherwise.
+			*/
+			template <typename K, typename Func>
+			bool find_with(K const& key, Func f)
+			{
+				return find_with(key, [=](K const& one, K const& two) { return one != two; }, f);
+			}
+
+			template <typename K, typename Predicate>
+			bool find(K const& key, Predicate pred)
+			{
+				return find_with(key, pred, [](mapped_type&) {});
+			}
+
+			template <typename K>
+			bool find(K const& key)
+			{
+				return find_with(key, [=](K const& one, K const& two) { return one != two; }, [](mapped_type&) {});
+			}
+
+			template <typename K, typename Predicate, typename Func>
+			bool find_with(K const& key, Predicate pred, Func f)
+			{
+				unsigned int hash = calc_hash(key);
```

Comment:
The hash function should probably return a `std::size_t` so that all indexes can be reached using the hash. If they are capped at the max int then a table whose size is `max int + 1` will never have those slots touched.

Reply:
@DaKellyFella, I actually already changed this in a local copy, haven't committed it yet. :smile: